### PR TITLE
Input model remain methods

### PR DESCRIPTION
**Pull request by itikhono**
### Details:
 - *implementation of remaining methods of InputModel class"

### Tickets:
 - *52259*

### CODE REVIEW

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 1]
```diff
@@ -502,6 +475,250 @@ namespace ngraph
...
+                traverse_down(m_inputs, nullptr, &new_tensors);
+                std::vector<Place::Ptr> new_outputs;
+                for (const auto& tensor : new_tensors)
+                {
+                    if (tensor->is_output())
+                    {
+                        FRONT_END_GENERAL_CHECK(std::find(target_out_tensor.begin(),
+                                                          target_out_tensor.end(),
+                                                          tensor) != target_out_tensor.end(),
+                                                "Incorrect Place for cutting.");
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/model.cpp:530]: Can it be more informative, i.e. add output's tensor's name which depends on new output?

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 2]
```diff
@@ -502,6 +475,250 @@ namespace ngraph
...
+            FRONT_END_GENERAL_CHECK(m_var_names.count(new_name) == 0,
+                                    "The provided name is already used in the model");
+            (*it)->add_name(new_name);
+            m_var_names[new_name] = *it;
+        }
+
+        void InputModelPDPD::InputModelPDPDImpl::setNameForTensor(Place::Ptr tensor,
+                                                                  const std::string& new_name)
+        {
+            auto it = std::find(m_var_places.begin(), m_var_places.end(), tensor);
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/model.cpp:593]: We're searching for place in the model, which has linear complexity
Instead, can 'Place' just check that it's 'm_inputModel' is the same as 'this'?

**Review by nosovmik — COMMENTED** [review]
Huge amount of code implemented, tests for this are extremely needed

File: ngraph/frontend/paddlepaddle/src/model.cpp (hunk 2 above)
Comment (itikhono) [reply ngraph/frontend/paddlepaddle/src/model.cpp:593]: I think so, good idea, thanks

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 3]
```diff
@@ -302,10 +281,10 @@ namespace ngraph
         void InputModelPDPD::InputModelPDPDImpl::loadConsts(
             const std::basic_string<T>& folder_with_weights, std::istream* weight_stream)
         {
-            for (const auto& item : m_var_places)
+            for (const auto& item : m_var_names)
             {
                 const auto& var_desc = item.second->get_desc();
-                const auto& name = item.first;
+                const auto& name = var_desc.name();
```
[the remark this answers is not in the document]
Comment (itikhono) [reply ngraph/frontend/paddlepaddle/src/model.cpp:287]: it's artifact of previous change, will be reverted

[the remark this answers is not in the document]
Comment (itikhono) [reply ngraph/frontend/paddlepaddle/src/model.cpp]: fixed

File: ngraph/frontend/paddlepaddle/src/model.cpp (hunk 2 above)
Comment (itikhono) [reply ngraph/frontend/paddlepaddle/src/model.cpp]: done

File: ngraph/frontend/paddlepaddle/src/model.cpp (hunk 1 above)
Comment (itikhono) [reply ngraph/frontend/paddlepaddle/src/model.cpp]: done

File: ngraph/frontend/paddlepaddle/src/pdpd_utils.cpp [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+                        q.push(op.get());
+                        if (ordered_ops)
+                            ordered_ops->push_back(op);
+                        return true;
+                    }
+                    return false;
+                };
+
+                for (const auto& node : start_nodes)
+                {
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/pdpd_utils.cpp:50]: I believe this cycle should just 'push to queue'. Attributes like 'visited' and ordered_ops will be filled in next cycle
Looks like there is some duplication here.
And also algorithm doesn't look like a topological sorting

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 5]
```diff
@@ -502,6 +446,140 @@ namespace ngraph
...
+            Place::Ptr place, const std::string& new_name_optional)
+        {
+            auto tensor_place = pdpd::castToTensorPlace(place);
+            if (tensor_place)
+            {
+                std::vector<std::shared_ptr<TensorPlacePDPD>> new_tensors;
+
+                // Find outputs that are connected to the cut place. These target outputs should be
+                // cut off.
+                pdpd::traverse_down({tensor_place}, nullptr, &new_tensors);
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/model.cpp:478]: When do we need second argument? Looks like it is always 'nullptr' and maybe be removed

Comment (itikhono) [reply ngraph/frontend/paddlepaddle/src/model.cpp:478]: we use it in `determine_cut_nodes()` function

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 6]
```diff
@@ -502,6 +446,140 @@ namespace ngraph
...
+                m_outputs.push_back(tensor_place);
+
+                // If some target outputs still connected with the model inputs, the selected cut
+                // place is incorrect.
+                new_tensors.clear();
+                pdpd::traverse_down(m_inputs, nullptr, &new_tensors);
+                std::vector<Place::Ptr> new_outputs;
+                for (const auto& tensor : new_tensors)
+                {
+                    if (tensor->is_output())
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/model.cpp:496]: Looks like a code duplication here and in loop body. I believe this duplication can be avoided

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 7]
```diff
@@ -502,6 +446,140 @@ namespace ngraph
...
+                    {
+                        target_out_tensor.push_back(tensor);
+                    }
+                }
+                m_outputs.push_back(tensor_place);
+
+                // If some target outputs still connected with the model inputs, the selected cut
+                // place is incorrect.
+                new_tensors.clear();
+                pdpd::traverse_down(m_inputs, nullptr, &new_tensors);
```
[the remark this answers is not in the document]
Comment (itikhono) [reply ngraph/frontend/paddlepaddle/src/model.cpp:492]: in fact we do just that, but we need to check first that the model is valid after cutting, this check is necessary
image

File: ngraph/frontend/paddlepaddle/src/model.cpp (hunk 6 above)
Comment (itikhono) [reply ngraph/frontend/paddlepaddle/src/model.cpp:496]: I guess it's ok, we have an additional check in the loop

File: ngraph/frontend/paddlepaddle/src/model.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (itikhono) [reply ngraph/frontend/paddlepaddle/src/model.cpp:492]: generic way is to pass callback to traverse function, and collect only necessary information
I will implement this approach if you do not mind

[the remark this answers is not in the document]
Comment (itikhono) [reply ngraph/frontend/paddlepaddle/src/model.cpp]: the traversals have been updated to use a callback

File: ngraph/frontend/paddlepaddle/src/pdpd_utils.cpp (hunk 4 above)
Comment (itikhono) [reply ngraph/frontend/paddlepaddle/src/pdpd_utils.cpp]: "also algorithm doesn't look like a topological sorting" - `traverse_up` is a topological sorting (tensor place has only one parent, input place is also has only one parent) 
`traverse_down` is not a topological sorting, but I believe it's unnecessary for our purpose

"I believe this cycle should just 'push to queue'. Attributes like 'visited' and ordered_ops will be filled in next cycle" - 
I guess we should do the same checks as in main loop

File: ngraph/frontend/paddlepaddle/src/model.cpp (hunk 6 above)
Comment (itikhono) [reply ngraph/frontend/paddlepaddle/src/model.cpp]: now we use callback

File: ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/place.hpp [hunk 8]
```diff
@@ -47,18 +48,26 @@ namespace ngraph
...
+            void set_name(const std::string& name) { m_names.assign({name}); }
+            void remove_name(const std::string& name)
+            {
+                m_names.erase(std::remove(m_names.begin(), m_names.end(), name), m_names.end());
+            }
 
-        private:
-            const InputModel& m_input_model;
+        protected:
+            const InputModelPDPD& m_input_model;
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/place.hpp:62]: By the way, why is it 'const reference' and not something like 'weak_ptr'?
I'm thinking about situations when model is destroyed by client, but Place object still exists. Like
```
Place::Ptr get_place_only() {
    InputModel::Ptr model = frontend.load("abc.pdmodel");
    return model.get_place_by_tensor_name("tensor");
}

auto place = get_place_only();
std::cout << place->is_input();
```
From API perspective, it is valid code and should not lead to crash. There can be an exception that model is destroyed, but not crash

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 9]
```diff
@@ -502,6 +453,142 @@ namespace ngraph
...
+        void InputModelPDPD::InputModelPDPDImpl::cutAndAddNewOutput(
+            Place::Ptr place, const std::string& new_name_optional)
+        {
+            auto tensor_place = pdpd::castToTensorPlace(place);
+            if (tensor_place)
+            {
+                // Find outputs that are connected to the cut place. These target outputs should be
+                // cut off.
+                std::vector<Place::Ptr> target_out_tensor;
+                auto callback_out = [&](const std::shared_ptr<OpPlacePDPD>& op_place,
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/model.cpp:484]: `op_place` is not used. Is it also 'reserved for future possible use'?

Comment (itikhono) [reply ngraph/frontend/paddlepaddle/src/model.cpp:484]: it's used in determine_cut_nodes() function

File: ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/place.hpp (hunk 8 above)
Comment (itikhono) [reply ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/place.hpp:62]: current pipeline: ctor InputModel -> ctor InputModelImpl (..., const InputModel&) -> LoadPlaces() -> Place(..., const InputModel&)
as I understood, you suggest
ctor InputModel -> ctor InputModelImpl (..., weak_ptr<InputModel>) -> LoadPlaces() -> Place(..., weak_ptr<InputModel>)

In this case, we need to inherit  InputModel from enable_shared_from_this and use shared_from_this in ctor
Unfortunately it's impossible to use shared_from_this() in ctors.

**Comment by ilya-lavrenov** [comment]
@itikhono is it still valid?

**Pull request closed without merging**